reason_4.0/lib/core/scripts/upgrade/4.7_to_4.8/add_event_registration.php

```diff
@@ -178,17 +199,146 @@ public function checkForEventSlots()
+	 * @return array
+	 */
+	public function getExistingSlotReport()
+	{
+		$type_exists = !empty(id_of("registration_slot_type"));
+		if (!$type_exists) {
+			return [];
+		}
+
+		$es = new entity_selector();
+		$es->add_type(id_of('event_type'));
+		$es->add_left_relationship_field('event_type_to_registration_slot_type', 'registration_slot', 'id', 'slot_id');
+		$events = $es->run_one();
+
+		$report = [];
+		foreach ($events as $event) {
+			$event_values = $event->get_values();
+
+			$slots = $event->get_left_relationship(relationship_id_of('event_type_to_registration_slot_type'));
+			foreach ($slots as $slot) {
+				$slot_values = $slot->get_values();
+
+				$event_info = array_combine(
+					array_map(function ($k) {
+						return 'event_' . $k;
+					}, array_keys($event_values)),
+					$event_values
+				);
+				$slot_info = array_combine(
+					array_map(function ($k) {
+						return 'slot_' . $k;
+					}, array_keys($slot_values)),
+					$slot_values
+				);
+				$report[] = array_merge($event_info, $slot_info);
+			}
+		}
+
+		return $report;
+	}
+
+	/**
+	 * Write a CSV to output buffers of existing registration slots
+	 */
 	public function exportOldSlots()
 	{
-		
+
+		$rows = $this->getExistingSlotReport();
+
+		if($rows) {
+			$filename = REASON_HOST . "_event_registration_slots_backup_" . date("Y-m-d-H-i-s") . ".csv";
+
+			ob_clean();
+			header("Content-Type: text/csv");
+			header('Content-Disposition: attachment; filename="' . $filename . '"');
+			$out = fopen('php://output', 'w');
+			fputcsv($out, array_keys($rows[0]));
+			foreach ($rows as $row) {
+				fputcsv($out, $row);
+			}
+			fclose($out);
+		}
+	}
+
+	/**
+	 * Remove all traces of Registration Slot in database
+	 *
+	 * @return bool TRUE on a successful delete of primary type, FALSE otherwise
+	 */
+	public function deleteEventSlots()
+	{
+		$type_exists = !empty(id_of("registration_slot_type"));
+
+		// Delete instances of Registration Slots
+		if ($type_exists) {
+			$es = new entity_selector();
+			$es->add_type(id_of('registration_slot_type'));
+			$slots = $es->run_one();
+			foreach ($slots as $slot) {
+				reason_expunge_entity($slot->id(), $this->user_id);
+			}
+		}
+
+		// Delete Registration Slot Entity Fields
+		if ($type_exists) {
+			$es = new entity_selector();
+			$es->add_type(id_of('field'));
+			$list = implode("', '", get_fields_by_content_table("registration_slot"));
+			$es->add_relation("`entity`.`name` IN ('$list')");
+			$fields = $es->run_one();
+			foreach ($fields as $field) {
+				reason_expunge_entity($field->id(), $this->user_id);
+			}
+		}
+
+		// Delete Registration Slot Entity Table
+		if ($type_exists) {
+			$es = new entity_selector();
+			$es->add_type(id_of('content_table'));
+			$es->add_relation("`entity`.`name` = 'registration_slot'");
+			$table = $es->run_one();
+			reason_expunge_entity(array_pop($table)->id(), $this->user_id);
+		}
+
+		// Delete Registration Slot Type
```

Comment:
Yeah, that'd be a good idea in retrospect. Since I know the current order of operations works, I'll probably leave it as is, though.